streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -0,0 +1,178 @@
+        stateRestorers.put(restorer.partition(), restorer);
+    }
+
+    public void restore() {
+        final long start = time.milliseconds();
+        try {
+            if (!consumer.subscription().isEmpty()) {
+                throw new IllegalStateException(String.format("Restore consumer should have not subscribed to any partitions (%s) beforehand", consumer.subscription()));
+            }
+            final Map<TopicPartition, Long> endOffsets = consumer.endOffsets(stateRestorers.keySet());
+            consumer.assign(stateRestorers.keySet());
+
+            seek();
+
+            final Set<TopicPartition> partitions = new HashSet<>(stateRestorers.keySet());
+            while (!partitions.isEmpty()) {
+                final ConsumerRecords<byte[], byte[]> allRecords = consumer.poll(10);
+                final Iterator<TopicPartition> partitionIterator = partitions.iterator();
+                while (partitionIterator.hasNext()) {
+                    restorePartition(endOffsets, allRecords, partitionIterator);
+                }
+            }
+        } finally {
+            consumer.assign(Collections.<TopicPartition>emptyList());
+            log.debug("Took {} ms to restore active state", time.milliseconds() - start);
+        }
+    }
+
+    private void restorePartition(final Map<TopicPartition, Long> endOffsets,
+                                  final ConsumerRecords<byte[], byte[]> allRecords,
+                                  final Iterator<TopicPartition> partitionIterator) {
+        final TopicPartition topicPartition = partitionIterator.next();
+        final StateRestorer restorer = stateRestorers.get(topicPartition);
+        final Long endOffset = endOffsets.get(topicPartition);
+        final long pos = processNext(allRecords.records(topicPartition), restorer);
+        if (restorer.hasCompleted(pos, endOffset)) {
+            if (pos > endOffset + 1) {
+                throw new IllegalStateException(
+                        String.format("Log end offset of %s should not change while restoring: old end offset %d, current offset %d",
+                                      topicPartition,
+                                      endOffset,
+                                      pos));
+            }
+            restorer.setRestoredOffset(pos);
+            partitionIterator.remove();
+        }
+    }
+
+    private void seek() {
+        for (final StateRestorer restorer : stateRestorers.values()) {
+            if (restorer.checkpoint() != null) {
+                consumer.seek(restorer.partition(), restorer.checkpoint());
+            } else {
+                consumer.seekToBeginning(Collections.singletonList(restorer.partition()));
+            }
+        }
+    }
+
+    private long processNext(final List<ConsumerRecord<byte[], byte[]>> records, final StateRestorer restorer) {
+        for (final ConsumerRecord<byte[], byte[]> record : records) {
+            final long offset = record.offset();
```

Comment:
I feel we have a duplicate logic here and in `StateRestorer`, since the latter has already a `readTo` function called by `hasCompleted` that already has considered offsetLimit. So to dedup the logic we can just call `restorer.restorer.hasCompleted(pos, endOffset)` here in line 152, and remove the additional check above.